finesse/gui/measure_script/script_edit_dialog.py

```diff
@@ -0,0 +1,437 @@
+
+        return super().headerData(section, orientation, role)
+
+    def rowCount(self, *args: Any, **kwargs: Any) -> int:
+        """Get the number of data rows."""
+        return len(self._sequence)
+
+    def columnCount(self, *args: Any, **kwargs: Any) -> int:
+        """Get the number of data columns."""
+        return 2
+
+
+class SequenceWidget(QWidget):
+    """A widget with a table of measure instructions and controls to modify them."""
+
+    def __init__(self, sequence: Optional[List[Measurement]] = None) -> None:
+        """Create a new SequenceWidget."""
+        super().__init__()
+
+        self.table = QTableView()
+
+        self.sequence = sequence if sequence else []
+        """The sequence of measure instructions as a list of angles and counts."""
+
+        self.model = SequenceModel(self.sequence)
+        self.table.setModel(self.model)
+        self.table.horizontalHeader().setStretchLastSection(True)
+        self.table.setSelectionBehavior(QAbstractItemView.SelectionBehavior.SelectRows)
+
+        self.buttons = SequenceButtons(self)
+
+        # Put the table on the left and the buttons on the right
+        layout = QHBoxLayout()
+        layout.addWidget(self.table)
+        layout.addWidget(self.buttons)
+        self.setLayout(layout)
+
+    def add_instruction(self, angle: Union[str, float], measurements: int) -> None:
+        """Add a new measure instruction to the sequence.
+
+        Args:
+            angle: Target angle as a float or a string corresponding to a preset
+            measurements: Number of times to take a measurement at this angle
+        """
+        self.sequence.append(Measurement(angle, measurements))
+        self.model.layoutChanged.emit()  # type: ignore
+        self.table.scrollToBottom()
+
+    def _get_selected_rows(self, reverse: bool = False) -> List[int]:
+        """Get the indices of the currently selected rows in the table.
+
+        Args:
+            reverse: Whether to return the indices in reverse order
+        """
+        selected = [idx.row() for idx in self.table.selectionModel().selectedRows()]
+        selected.sort(reverse=reverse)
+        return selected
+
+    def _swap_rows(self, lower: int, higher: int) -> None:
+        """Swap the rows specified by two indices."""
+        # For some reason if these arguments are the wrong way round, Qt crashes
+        assert higher > lower
+
+        self.model.beginMoveRows(QModelIndex(), higher, higher, QModelIndex(), lower)
+        self.sequence[lower], self.sequence[higher] = (
+            self.sequence[higher],
+            self.sequence[lower],
+        )
+        self.model.endMoveRows()
+
+    def move_selected_up(self) -> None:
+        """Move the currently selected instructions up one row."""
+        selected = self._get_selected_rows()
+        if not selected or selected[0] == 0:
+            # We can't move up if the topmost item is selected
```

Comment:
I wonder if it's best to have these buttons disabled unless a row is selected?

Reply:
Yeah, they really should be. There are actually a bunch of places on the GUI where we should be disabling buttons when they're not usable and aren't.

I'll maybe look at this when I write the tests for this lot.